kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java

```diff
@@ -250,43 +261,61 @@ private boolean checkRotationOrAppend(
       String encodedPartition,
       long now
   ) {
-    if (compatibility.shouldChangeSchema(record, null, currentValueSchema)
-        && recordCount > 0) {
-      // This branch is never true for the first record read by this TopicPartitionWriter
-      log.trace(
-          "Incompatible change of schema detected for record '{}' with encoded partition "
-          + "'{}' and current offset: '{}'",
-          record,
-          encodedPartition,
-          currentOffset
-      );
-      currentSchemas.put(encodedPartition, valueSchema);
-      nextState();
-    } else if (rotateOnTime(encodedPartition, currentTimestamp, now)) {
-      setNextScheduledRotation();
-      nextState();
-    } else {
-      currentEncodedPartition = encodedPartition;
-      SinkRecord projectedRecord = compatibility.project(
-          record,
-          null,
-          currentValueSchema
-      );
-      writeRecord(projectedRecord);
-      buffer.poll();
-      if (rotateOnSize()) {
-        log.info(
-            "Starting commit and rotation for topic partition {} with start offset {}",
-            tp,
-            startOffsets
+    try {
+      if (compatibility.shouldChangeSchema(record, null, currentValueSchema)
+          && recordCount > 0) {
+        // This branch is never true for the first record read by this TopicPartitionWriter
+        log.trace(
+            "Incompatible change of schema detected for record '{}' with encoded partition "
+                + "'{}' and current offset: '{}'",
+            record,
+            encodedPartition,
+            currentOffset
         );
+        currentSchemas.put(encodedPartition, valueSchema);
+        nextState();
+      } else if (rotateOnTime(encodedPartition, currentTimestamp, now)) {
+        setNextScheduledRotation();
         nextState();
-        // Fall through and try to rotate immediately
       } else {
-        return false;
+        currentEncodedPartition = encodedPartition;
+        SinkRecord projectedRecord = compatibility.project(
+            record,
+            null,
+            currentValueSchema
+        );
+        writeRecord(projectedRecord);
+        buffer.poll();
+        if (rotateOnSize()) {
+          log.info(
+              "Starting commit and rotation for topic partition {} with start offset {}",
+              tp,
+              startOffsets
+          );
+          nextState();
+          // Fall through and try to rotate immediately
+        } else {
+          return false;
+        }
+      }
+      return true;
+    } catch (SchemaProjectorException e) {
+      if (reporter != null) {
+        reporter.report(record, e);
+        log.warn("Errant record written to DLQ due to SchemaProjectorException.");
+        buffer.poll();
```

Comment:
The tests using mock S3 fail if this buffer.poll() is not present. They fail on attempts to write records after the errant record with the exception: 
```
org.apache.avro.AvroRuntimeException: already open

	at org.apache.avro.file.DataFileWriter.assertNotOpen(DataFileWriter.java:93)
	at org.apache.avro.file.DataFileWriter.setCodec(DataFileWriter.java:101)
	at io.confluent.connect.s3.format.avro.AvroRecordWriterProvider$1.write(AvroRecordWriterProvider.java:75)
	at io.confluent.connect.s3.TopicPartitionWriter.writeRecord(TopicPartitionWriter.java:530)
	at io.confluent.connect.s3.TopicPartitionWriter.checkRotationOrAppend(TopicPartitionWriter.java:287)
	at io.confluent.connect.s3.TopicPartitionWriter.executeState(TopicPartitionWriter.java:231)
	at io.confluent.connect.s3.TopicPartitionWriter.write(TopicPartitionWriter.java:200)
	at io.confluent.connect.s3.TopicPartitionWriterTest.testSchemaProjectionExceptionReported(TopicPartitionWriterTest.java:418)
```

Reply:
correct